src/OpenFeature/Api.cs

```diff
@@ -36,31 +35,26 @@ public sealed class Api
         private Api() { }
 
         /// <summary>
-        /// Sets the feature provider
+        /// Sets the feature provider. In order to wait for the provider to be set, and initialization to complete,
+        /// await the returned task.
         /// </summary>
+        /// <remarks>The provider cannot be set to null. Attempting to set the provider to null has no effect.</remarks>
         /// <param name="featureProvider">Implementation of <see cref="FeatureProvider"/></param>
-        public void SetProvider(FeatureProvider featureProvider)
+        public async Task SetProvider(FeatureProvider featureProvider)
         {
-            this._featureProviderLock.EnterWriteLock();
-            try
-            {
-                this._defaultProvider = featureProvider ?? this._defaultProvider;
-            }
-            finally
-            {
-                this._featureProviderLock.ExitWriteLock();
-            }
+            await this._repository.SetProvider(featureProvider, this.GetContext()).ConfigureAwait(false);

```

Comment:
I refactored the collection management out of the Api and moved it into the ProviderRepository. This is similar to the Java SDK.